compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/KernelDispatch.cpp

```diff
@@ -17,6 +18,36 @@
 
 namespace mlir::iree_compiler::AMDAIE {
 
+// Find the largest factor of 'num' which is not larger than 'max'.
+int detail::findLargestFactor(int num, int max) {
+  assert(max > 0 && "No factors less than or equal to 0 exist");
+
+  // Do O(1) instead of O(sqrt(num)) computation for this common case.
+  if (num <= max) {
+    return num;
+  }
+
+  int largestLowFactor = 1;
+  for (int lowFactor = 2; lowFactor <= max; ++lowFactor) {
+    const int highFactor = num / lowFactor;
+
+    // This early exit is what makes this O(sqrt(num)) instead of O(num).
+    if (highFactor < lowFactor) return largestLowFactor;
+
+    const bool areActuallyFactors = num % lowFactor == 0;
+    if (areActuallyFactors) {
```

Comment:
I know @nirvedhmeshram is the original author of this function, but I think it's better to move this utility to AMDAIEUtils.cpp.

Reply:
I've moved it. FYI the only reason I've put this in a header at all is so that it can be tested. Because it's not really intended for end user use, I've put it in the 'detail' namespace. This seems to a be a common pattern in LLVM from what I can tell